akka-distributed-data/src/main/scala/akka/cluster/ddata/Replicator.scala

```diff
@@ -0,0 +1,1577 @@
+  val selfAddress = cluster.selfAddress
+  val selfUniqueAddress = cluster.selfUniqueAddress
+
+  require(!cluster.isTerminated, "Cluster node must not be terminated")
+  require(role.forall(cluster.selfRoles.contains),
+    s"This cluster member [${selfAddress}] doesn't have the role [$role]")
+
+  //Start periodic gossip to random nodes in cluster
+  import context.dispatcher
+  val gossipTask = context.system.scheduler.schedule(gossipInterval, gossipInterval, self, GossipTick)
+  val notifyTask = context.system.scheduler.schedule(notifySubscribersInterval, notifySubscribersInterval, self, FlushChanges)
+  val pruningTask = context.system.scheduler.schedule(pruningInterval, pruningInterval, self, RemovedNodePruningTick)
+  val clockTask = context.system.scheduler.schedule(gossipInterval, gossipInterval, self, ClockTick)
+
+  val serializer = SerializationExtension(context.system).serializerFor(classOf[DataEnvelope])
+  val maxPruningDisseminationNanos = maxPruningDissemination.toNanos
+
+  // cluster nodes, doesn't contain selfAddress
+  var nodes: Set[Address] = Set.empty
+
+  // nodes removed from cluster, to be pruned, and tombstoned
+  var removedNodes: Map[UniqueAddress, Long] = Map.empty
+  var pruningPerformed: Map[UniqueAddress, Long] = Map.empty
+  var tombstoneNodes: Set[UniqueAddress] = Set.empty
+
+  var leader: Option[Address] = None
+  def isLeader: Boolean = leader.exists(_ == selfAddress)
+
+  // for pruning timeouts are based on clock that is only increased when all nodes are reachable
+  var previousClockTime = System.nanoTime()
+  var allReachableClockTime = 0L
+  var unreachable = Set.empty[Address]
+
+  // the actual data
+  var dataEntries = Map.empty[String, (DataEnvelope, Digest)]
+  var changed = Map.empty[String, Digest]
+
+  // for splitting up gossip in chunks
+  var statusCount = 0L
+  var statusTotChunks = 0
+
+  val subscribers = new mutable.HashMap[String, mutable.Set[ActorRef]] with mutable.MultiMap[String, ActorRef]
+  val newSubscribers = new mutable.HashMap[String, mutable.Set[ActorRef]] with mutable.MultiMap[String, ActorRef]
+
+  var updateInProgressBuffer = Map.empty[String, Queue[BufferedCommand]]
+
+  override def preStart(): Unit = {
+    val leaderChangedClass = if (role.isDefined) classOf[RoleLeaderChanged] else classOf[LeaderChanged]
+    cluster.subscribe(self, initialStateMode = InitialStateAsEvents,
+      classOf[MemberEvent], classOf[ReachabilityEvent], leaderChangedClass)
+  }
+
+  override def postStop(): Unit = {
+    cluster.unsubscribe(self)
+    gossipTask.cancel()
+    notifyTask.cancel()
+    pruningTask.cancel()
+    clockTask.cancel()
+  }
+
+  def matchingRole(m: Member): Boolean = role.forall(m.hasRole)
+
+  def receive = normalReceive
+
+  val normalReceive: Receive = {
+    case Get(key, consistency, req)                 ⇒ receiveGet(key, consistency, req, sender())
+    case Update(key, _, _, req) if !isLocalSender() ⇒ receiveInvalidUpdate(key, req)
```

Comment:
but on the other hand, the function will not accidentally make it through serialization for such cases. I removed this, in case someone cares to write a fancy serializer.